test/shared/src/main/scala/zio/test/environment/package.scala

```diff
@@ -335,83 +338,156 @@ package object environment extends PlatformSpecific {
-       * time is adjusted to on or after the duration, the fiber will
-       * automatically be resumed.
+       * Semantically blocks the current fiber until the clock time is equal
+       * to or greater than the specified duration. Once the clock time is
+       * adjusted to on or after the duration, the fiber will automatically be
+       * resumed.
        */
       def sleep(duration: Duration): UIO[Unit] =
         for {
-          latch <- Promise.make[Nothing, Unit]
+          fiberId <- ZIO.fiberId
+          latch   <- Promise.make[Nothing, Unit]
           start <- fiberState.modify { data =>
                     val end = data.duration + duration
-                    (data.duration, FiberData(end, data.timeZone))
+                    (data.duration, FiberData(end))
                   }
-          await <- clockState.modify { data =>
+          await <- clockState.get.flatMap { data =>
                     val end = start + duration
                     if (end > data.duration)
-                      (true, data.copy(sleeps = (end, latch) :: data.sleeps))
+                      scheduled.offer(end, Sleep(end, latch, fiberId)).as(true)
                     else
-                      (false, data)
-                  }
+                      STM.succeedNow(false)
+                  }.commit
           _ <- if (await) warningStart *> latch.await else latch.succeed(())
         } yield ()
 
       /**
-       * Returns a list of the wall clock times at which all queued effects are
-       * scheduled to resume.
+       * Returns a list of the times at which all queued effects are scheduled
+       * to resume.
        */
       lazy val sleeps: UIO[List[Duration]] =
-        clockState.get.map(_.sleeps.map(_._1))
+        scheduled.toList.map(_.map(_.duration)).commit
 
       /**
        * Returns the time zone.
        */
       lazy val timeZone: UIO[ZoneId] =
-        fiberState.get.map(_.timeZone)
-
-      private def run(wakes: List[(Duration, Promise[Nothing, Unit])]): UIO[Unit] =
-        UIO.foreach(wakes.sortBy(_._1))(_._2.succeed(())).unit
+        clockState.get.map(_.timeZone).commit
 
+      /**
+       * Cancels the warning message that is displayed if a test is using time
+       * but is not advancing the `TestClock`.
+       */
       private[TestClock] val warningDone: UIO[Unit] =
         warningState.updateSome[Any, Nothing] {
           case WarningData.Start          => ZIO.succeedNow(WarningData.done)
           case WarningData.Pending(fiber) => fiber.interrupt.as(WarningData.done)
         }
 
+      /**
+       * Polls untl the specified fiber completes execution or is indefinitely
+       * suspended, meaning that the fiber and all the fibers it is blocking on
+       * and all the fibers they are blocking on recursively are suspended.
+       */
+      private def awaitSuspended(fiberId: Fiber.Id): UIO[Unit] =
+        live.provide {
+          suspended(fiberId).repeat {
```

Comment:
As `Fiber.get` can be a rather expensive operation if we have many Fibers, would it make sense to split this logic into a first step, which just gets the Fiber by id and a second which checks for suspension? Suspended could be modified to take a Fiber instead of just an Id.

Reply:
Yes I think some kind of memoization of the fibers we get definitely makes sense if we can't just look them up directly.